>### **Project Title:** Electronic Shop Sales Analysis
>### **Project Owner:** Berlinda Anaman
>### **Email:** Berlana.d@gmail.com
>### **Github Profile:** [Berl-cloud](https://github.com/Berl-cloud)
>### **LinkedIn Profile:** [Berlinda Anaman](http://www.linkedin.com/in/berlinda-anaman)

# Introduction
Welcome to the exploratory data analysis of sales data of an electronics shop specializing in products such as laptops, phones, fridges etc. The electronics industry is dynamic and rapidly evolving, with new products and technologies constantly entering the market. For an electronics shop to thrive in this competitive landscape, it's crucial to gain insights into customer preferences, product demand, and sales patterns. In this notebook, we will dive into the dataset to understand sales trends, customer behavior, and product performance. We aim to uncover actionable insights that can inform strategic decisions and drive business growth.


# Project Scope

The objective of this project is to dive into the large sales dataset to extract valuable insights. We will explore sales trends over time, identify the best-selling products, calculate revenue metrics such as total sales and profit margins, and create visualizations to present our findings effectively. We will use python in exploring and cleaning the data, and use Power BI to build a dashboard of the cleaned data.

## 1. Data Preparation

An understanding of the data coupled with problem understanding will help us in cleaning and preparing our data for analysis. It is usually rare to acquire a ready-to-use data for any analysis without some level of preparation. To prepare our data, we normally assess the quality of the data, cleanse, format, blend and sample the data since we may encounter various issues with columns in our data. These issues may include:

* **`Missing values:`** meaning column values are incomplete
* **`Incorrect data:`** meaning you see values not expected for the column name
* **`Inconsistent values:`** meaning some values may fall outside the expected range
* **`Duplicate values:`** meaning whether or not there are duplicate values
* **`Inconsistent data type:`** meaning values entered in the columns may not be consistent with the column names

To properly prepare our data for analysis, we will perform two important tasks which are;

* Part I: Data Quality Assessment
* Part II: Data Cleaning and Preprocessing 


### 1.1 Prerequisites

Before we proceed to Part I of the data preparation , we will need to explore the dataset to understand its structure. We will first of all import the necessary libraries that will be used in the project. In this case, we need pandas and numpy. 

Pandas is a library used for data manipulation and analysis. Numpy is a library used for numerical computations.

In [2]:
#Import required libaries 

import pandas as pd
import numpy as np
from datetime import datetime

In [3]:
# load sales data

data = pd.read_csv("Data\Sales Data.csv")


In [4]:
# Load the data into a Pandas DataFrame

df = pd.DataFrame(data)

In [5]:
df.head() #inspect the data

Unnamed: 0  Order ID               Product  Quantity Ordered  Price Each  \
0           0    295665    Macbook Pro Laptop                 1     1700.00   
1           1    295666    LG Washing Machine                 1      600.00   
2           2    295667  USB-C Charging Cable                 1       11.95   
3           3    295668      27in FHD Monitor                 1      149.99   
4           4    295669  USB-C Charging Cable                 1       11.95   

         Order Date                        Purchase Address  Month    Sales  \
0   12/30/2019 0:01  136 Church St, New York City, NY 10001     12  1700.00   
1   12/29/2019 7:03     562 2nd St, New York City, NY 10001     12   600.00   
2  12/12/2019 18:21    277 Main St, New York City, NY 10001     12    11.95   
3  12/22/2019 15:13     410 6th St, San Francisco, CA 94016     12   149.99   
4  12/18/2019 12:38           43 Hill St, Atlanta, GA 30301     12    11.95   

             City  Hour  
0   New York City     0  
1   New York City     7  
2   New York City    18  
3   San Francisco    15  
4         Atlanta    12

### 1.2 Data Exploration

In this step, we will familiarise ourselves with the data. We will have an overview of the data by printing the first few rows to see how the data looks like. 
 
We will also check some informations about the data such as the various columns, their data type as well as the number of missing values. 

We wil then view a summary of the statistics of the data such as the number of observations. 
 
Lastly, we will print the number of observations for each category of the social media engagements.

In [6]:
# A glimpse at the data 

df.head()

Unnamed: 0  Order ID               Product  Quantity Ordered  Price Each  \
0           0    295665    Macbook Pro Laptop                 1     1700.00   
1           1    295666    LG Washing Machine                 1      600.00   
2           2    295667  USB-C Charging Cable                 1       11.95   
3           3    295668      27in FHD Monitor                 1      149.99   
4           4    295669  USB-C Charging Cable                 1       11.95   

         Order Date                        Purchase Address  Month    Sales  \
0   12/30/2019 0:01  136 Church St, New York City, NY 10001     12  1700.00   
1   12/29/2019 7:03     562 2nd St, New York City, NY 10001     12   600.00   
2  12/12/2019 18:21    277 Main St, New York City, NY 10001     12    11.95   
3  12/22/2019 15:13     410 6th St, San Francisco, CA 94016     12   149.99   
4  12/18/2019 12:38           43 Hill St, Atlanta, GA 30301     12    11.95   

             City  Hour  
0   New York City     0  
1   New York City     7  
2   New York City    18  
3   San Francisco    15  
4         Atlanta    12

In [7]:
df.shape

(185950, 11)

In [8]:
# View data information

df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 185950 entries, 0 to 185949
Data columns (total 11 columns):
 #   Column            Non-Null Count   Dtype  
---  ------            --------------   -----  
 0   Unnamed: 0        185950 non-null  int64  
 1   Order ID          185950 non-null  int64  
 2   Product           185950 non-null  object 
 3   Quantity Ordered  185950 non-null  int64  
 4   Price Each        185950 non-null  float64
 5   Order Date        185950 non-null  object 
 6   Purchase Address  185950 non-null  object 
 7   Month             185950 non-null  int64  
 8   Sales             185950 non-null  float64
 9   City              185950 non-null  object 
 10  Hour              185950 non-null  int64  
dtypes: float64(2), int64(5), object(4)
memory usage: 15.6+ MB


In [9]:
# View summary statistics of the data

df.describe(include='all')

Unnamed: 0       Order ID               Product  Quantity Ordered  \
count   185950.000000  185950.000000                185950     185950.000000   
unique            NaN            NaN                    19               NaN   
top               NaN            NaN  USB-C Charging Cable               NaN   
freq              NaN            NaN                 21903               NaN   
mean      8340.388475  230417.569379                   NaN          1.124383   
std       5450.554093   51512.737110                   NaN          0.442793   
min          0.000000  141234.000000                   NaN          1.000000   
25%       3894.000000  185831.250000                   NaN          1.000000   
50%       7786.000000  230367.500000                   NaN          1.000000   
75%      11872.000000  275035.750000                   NaN          1.000000   
max      25116.000000  319670.000000                   NaN          9.000000   

           Price Each        Order Date  \
count   185950.000000            185950   
unique            NaN            142395   
top               NaN  12/15/2019 20:16   
freq              NaN                 8   
mean       184.399735               NaN   
std        332.731330               NaN   
min          2.990000               NaN   
25%         11.950000               NaN   
50%         14.950000               NaN   
75%        150.000000               NaN   
max       1700.000000               NaN   

                              Purchase Address          Month          Sales  \
count                                   185950  185950.000000  185950.000000   
unique                                  140787            NaN            NaN   
top     193 Forest St, San Francisco, CA 94016            NaN            NaN   
freq                                         9            NaN            NaN   
mean                                       NaN       7.059140     185.490917   
std                                        NaN       3.502996     332.919771   
min                                        NaN       1.000000       2.990000   
25%                                        NaN       4.000000      11.950000   
50%                                        NaN       7.000000      14.950000   
75%                                        NaN      10.000000     150.000000   
max                                        NaN      12.000000    3400.000000   

                  City           Hour  
count           185950  185950.000000  
unique               9            NaN  
top      San Francisco            NaN  
freq             44732            NaN  
mean               NaN      14.413305  
std                NaN       5.423416  
min                NaN       0.000000  
25%                NaN      11.000000  
50%                NaN      15.000000  
75%                NaN      19.000000  
max                NaN      23.000000

## Part I: Data Quality Assessment

Here, we will assess the quality of the data to know if it is ready for the analyses. We will do this by inspecting for missing values, duplicated values, wrong data types, outliers and so on.

This task will enable us to determine how to clean the data and what sort of cleaning is needed.

In [10]:
# checking for missing values

df.isna().sum()

Unnamed: 0          0
Order ID            0
Product             0
Quantity Ordered    0
Price Each          0
Order Date          0
Purchase Address    0
Month               0
Sales               0
City                0
Hour                0
dtype: int64

In [11]:
# checking for duplicatd values

df.duplicated().sum()

0

In [12]:
df.dtypes

Unnamed: 0            int64
Order ID              int64
Product              object
Quantity Ordered      int64
Price Each          float64
Order Date           object
Purchase Address     object
Month                 int64
Sales               float64
City                 object
Hour                  int64
dtype: object

### Data Quality Assessment Report

This is the report of the above data quality assessment. The outcomes are outlined below:

1. It is clear that, there are no missing values in the data. All the data fields are filled with their corresponding values.
2. The data has no duplicate values. All observations are unique.
3. `Unnamed` column is not needed so will be deleted. Any unwanted column will be deleted.
4. `Order Date` column has a data type of object so we will format it to date data type. Also we will split the order date column into date, day and month.
5. We need to format `Month` column by converting the numerical representation of the month to text values.

**Based on the data quality assessment report above, we will go ahead with the necessary data cleaning.** 

## Part II: Data Cleaning and Preprocessing

During the data quality assessment, we identified few data quality issues including unwanted data, inconsistent values, etc. 

In this task we will perform all the initial data cleaning and preprocessing needed to produce data that will be suitable for our analysis.</p>

In [13]:
# dropping unneeded column

df = df.drop(columns=['Unnamed: 0'])


In [14]:
df.head()

Order ID               Product  Quantity Ordered  Price Each  \
0    295665    Macbook Pro Laptop                 1     1700.00   
1    295666    LG Washing Machine                 1      600.00   
2    295667  USB-C Charging Cable                 1       11.95   
3    295668      27in FHD Monitor                 1      149.99   
4    295669  USB-C Charging Cable                 1       11.95   

         Order Date                        Purchase Address  Month    Sales  \
0   12/30/2019 0:01  136 Church St, New York City, NY 10001     12  1700.00   
1   12/29/2019 7:03     562 2nd St, New York City, NY 10001     12   600.00   
2  12/12/2019 18:21    277 Main St, New York City, NY 10001     12    11.95   
3  12/22/2019 15:13     410 6th St, San Francisco, CA 94016     12   149.99   
4  12/18/2019 12:38           43 Hill St, Atlanta, GA 30301     12    11.95   

             City  Hour  
0   New York City     0  
1   New York City     7  
2   New York City    18  
3   San Francisco    15  
4         Atlanta    12

In [15]:
#format data type

df["Order Date"] = df["Order Date"].astype("datetime64[ns]")

In [16]:
df.dtypes

Order ID                     int64
Product                     object
Quantity Ordered             int64
Price Each                 float64
Order Date          datetime64[ns]
Purchase Address            object
Month                        int64
Sales                      float64
City                        object
Hour                         int64
dtype: object

In [17]:
#split Order Date column into Date, Order year, Order Day, Order Month and Order Hour

df['Order Day'] = pd.to_datetime(df['Order Date']).dt.day

df['Order Month'] = pd.to_datetime(df['Order Date']).dt.month

df['Order Year'] = pd.to_datetime(df['Order Date']).dt.year

df['Order Hour'] = pd.to_datetime(df['Order Date']).dt.hour

df['Date'] = pd.to_datetime(df['Order Date']).dt.date 

In [18]:
#inspect the data

df.head()

Order ID               Product  Quantity Ordered  Price Each  \
0    295665    Macbook Pro Laptop                 1     1700.00   
1    295666    LG Washing Machine                 1      600.00   
2    295667  USB-C Charging Cable                 1       11.95   
3    295668      27in FHD Monitor                 1      149.99   
4    295669  USB-C Charging Cable                 1       11.95   

           Order Date                        Purchase Address  Month    Sales  \
0 2019-12-30 00:01:00  136 Church St, New York City, NY 10001     12  1700.00   
1 2019-12-29 07:03:00     562 2nd St, New York City, NY 10001     12   600.00   
2 2019-12-12 18:21:00    277 Main St, New York City, NY 10001     12    11.95   
3 2019-12-22 15:13:00     410 6th St, San Francisco, CA 94016     12   149.99   
4 2019-12-18 12:38:00           43 Hill St, Atlanta, GA 30301     12    11.95   

             City  Hour  Order Day  Order Month  Order Year  Order Hour  \
0   New York City     0         30           12        2019           0   
1   New York City     7         29           12        2019           7   
2   New York City    18         12           12        2019          18   
3   San Francisco    15         22           12        2019          15   
4         Atlanta    12         18           12        2019          12   

         Date  
0  2019-12-30  
1  2019-12-29  
2  2019-12-12  
3  2019-12-22  
4  2019-12-18

**Now we will compare `Month` column with `Order Month` and `Hour` column with `Order Hour` to know their similarities.**

In [19]:
# comparing Month and Order Month columns

df['Month'].equals(df['Order Month'])

True

In [20]:
# comparing Hour and Order Hour columns

df['Hour'].equals(df['Order Hour'])

True

> **The output indicates that both `Month` and `Order Month` columns are the same as well as `Hour` and `Order Hour` columns. Hence we will delete one of each. For readability, we will delete `Oder Date` column which is no longer needed, `Month` and `Order Hour`coulmns and rename `Order Day` column as `Day`.**

In [21]:
# dropping unneeded columns

df = df.drop(columns=['Order Hour', 'Month', 'Order Date'])

In [22]:
# rename Order Day column as Day and Order Year coulmn as Year

df.rename(columns={'Order Day':'Day','Order Month':'Month', 'Order Year':'Year'}, inplace = True)

In [23]:
df.head()

Order ID               Product  Quantity Ordered  Price Each  \
0    295665    Macbook Pro Laptop                 1     1700.00   
1    295666    LG Washing Machine                 1      600.00   
2    295667  USB-C Charging Cable                 1       11.95   
3    295668      27in FHD Monitor                 1      149.99   
4    295669  USB-C Charging Cable                 1       11.95   

                         Purchase Address    Sales            City  Hour  Day  \
0  136 Church St, New York City, NY 10001  1700.00   New York City     0   30   
1     562 2nd St, New York City, NY 10001   600.00   New York City     7   29   
2    277 Main St, New York City, NY 10001    11.95   New York City    18   12   
3     410 6th St, San Francisco, CA 94016   149.99   San Francisco    15   22   
4           43 Hill St, Atlanta, GA 30301    11.95         Atlanta    12   18   

   Month  Year        Date  
0     12  2019  2019-12-30  
1     12  2019  2019-12-29  
2     12  2019  2019-12-12  
3     12  2019  2019-12-22  
4     12  2019  2019-12-18

> **Now, we will convert the values of the `Month` column to text and split Order Date into date and time**

In [24]:
#convert numeric month values to text month values

df['Month'] = pd.to_datetime(df['Month'], format='%m').dt.month_name()

In [25]:
#inspecting the first few rows

df.head()

Order ID               Product  Quantity Ordered  Price Each  \
0    295665    Macbook Pro Laptop                 1     1700.00   
1    295666    LG Washing Machine                 1      600.00   
2    295667  USB-C Charging Cable                 1       11.95   
3    295668      27in FHD Monitor                 1      149.99   
4    295669  USB-C Charging Cable                 1       11.95   

                         Purchase Address    Sales            City  Hour  Day  \
0  136 Church St, New York City, NY 10001  1700.00   New York City     0   30   
1     562 2nd St, New York City, NY 10001   600.00   New York City     7   29   
2    277 Main St, New York City, NY 10001    11.95   New York City    18   12   
3     410 6th St, San Francisco, CA 94016   149.99   San Francisco    15   22   
4           43 Hill St, Atlanta, GA 30301    11.95         Atlanta    12   18   

      Month  Year        Date  
0  December  2019  2019-12-30  
1  December  2019  2019-12-29  
2  December  2019  2019-12-12  
3  December  2019  2019-12-22  
4  December  2019  2019-12-18

For the purpose of our analysis, we will convert the weekday data from integer to text.  

In [26]:
df['Weekday'] = pd.to_datetime(df['Date']).dt.strftime('%A')

In [27]:
df.head()

Order ID               Product  Quantity Ordered  Price Each  \
0    295665    Macbook Pro Laptop                 1     1700.00   
1    295666    LG Washing Machine                 1      600.00   
2    295667  USB-C Charging Cable                 1       11.95   
3    295668      27in FHD Monitor                 1      149.99   
4    295669  USB-C Charging Cable                 1       11.95   

                         Purchase Address    Sales            City  Hour  Day  \
0  136 Church St, New York City, NY 10001  1700.00   New York City     0   30   
1     562 2nd St, New York City, NY 10001   600.00   New York City     7   29   
2    277 Main St, New York City, NY 10001    11.95   New York City    18   12   
3     410 6th St, San Francisco, CA 94016   149.99   San Francisco    15   22   
4           43 Hill St, Atlanta, GA 30301    11.95         Atlanta    12   18   

      Month  Year        Date    Weekday  
0  December  2019  2019-12-30     Monday  
1  December  2019  2019-12-29     Sunday  
2  December  2019  2019-12-12   Thursday  
3  December  2019  2019-12-22     Sunday  
4  December  2019  2019-12-18  Wednesday

**Now we can drop `Date column` peacefully.**

In [28]:
df = df.drop(columns=['Date'])

In [29]:
df.head()

Order ID               Product  Quantity Ordered  Price Each  \
0    295665    Macbook Pro Laptop                 1     1700.00   
1    295666    LG Washing Machine                 1      600.00   
2    295667  USB-C Charging Cable                 1       11.95   
3    295668      27in FHD Monitor                 1      149.99   
4    295669  USB-C Charging Cable                 1       11.95   

                         Purchase Address    Sales            City  Hour  Day  \
0  136 Church St, New York City, NY 10001  1700.00   New York City     0   30   
1     562 2nd St, New York City, NY 10001   600.00   New York City     7   29   
2    277 Main St, New York City, NY 10001    11.95   New York City    18   12   
3     410 6th St, San Francisco, CA 94016   149.99   San Francisco    15   22   
4           43 Hill St, Atlanta, GA 30301    11.95         Atlanta    12   18   

      Month  Year    Weekday  
0  December  2019     Monday  
1  December  2019     Sunday  
2  December  2019   Thursday  
3  December  2019     Sunday  
4  December  2019  Wednesday

**Our data looks clean and organised. At this point, we will move on to the next process.**

## 2. Exploratory Data Analysis and Visualisation

Here, we are going to perform statistical analysis on the data to answer a few questions and also create visualisations to get a pictorial undrstanding of the data.

Questions to be answered include:

1. What is the total sales in the year 2019 and 2020.

2. Which month in the year recorded the highest amount of sales.

3. Which product has the highest/lowest price. 
 
4. Which product was the most/least purchased in each year.

5. What is the total quantity items purchased. 

6. Which city recorded the most/least sales. 

7. Which day of the week has the most/least sales.

8. In which hour of the day do we experience more sales.

## 3. Statistical Analysis

#### 1. Total Sales

In [30]:
# Printing total sales made

total_sales = df['Sales'].sum()

yearly_sales = df.groupby('Year')['Sales'].sum()

In [31]:
print('Overall Sales:', total_sales)

Overall Sales: 34492035.97


In [32]:
print(pd.DataFrame(yearly_sales))

            Sales
Year             
2019  34483365.68
2020      8670.29


> Sales for the year 2020 is too little, so we will explore that to understand why.

#### Breakdown of 2020 sales recorded

In [33]:
#Data from only year 2020

df[df['Year'] == 2020]

Order ID                     Product  Quantity Ordered  Price Each  \
1555     297150    Lightning Charging Cable                 1       14.95   
2255     297817                      iPhone                 1      700.00   
2256     297817    Lightning Charging Cable                 2       14.95   
2415     297969                Google Phone                 1      600.00   
2796     298344            Wired Headphones                 4       11.99   
3537     299049    Apple Airpods Headphones                 1      150.00   
3617     299125        USB-C Charging Cable                 1       11.95   
4914     300362        USB-C Charging Cable                 1       11.95   
5076     300519  Bose SoundSport Headphones                 1       99.99   
5077     300519    Lightning Charging Cable                 1       14.95   
5326     300759                20in Monitor                 1      109.99   
7012     302393    Lightning Charging Cable                 1       14.95   
8867     304165      AAA Batteries (4-pack)                 1        2.99   
10301    305538    Apple Airpods Headphones                 1      150.00   
10614    305840  Bose SoundSport Headphones                 1       99.99   
11463    306655             ThinkPad Laptop                 1      999.99   
11625    306809          Macbook Pro Laptop                 1     1700.00   
11751    306928             ThinkPad Laptop                 1      999.99   
12739    307890    Lightning Charging Cable                 1       14.95   
13046    308185            27in FHD Monitor                 1      149.99   
13187    308322  Bose SoundSport Headphones                 1       99.99   
15009    310074    Apple Airpods Headphones                 1      150.00   
16364    311386                      iPhone                 1      700.00   
16365    311386    Apple Airpods Headphones                 1      150.00   
17018    312012            27in FHD Monitor                 1      149.99   
17855    312812               Flatscreen TV                 1      300.00   
19166    314066        USB-C Charging Cable                 1       11.95   
19666    314552      AAA Batteries (4-pack)                 1        2.99   
20276    315138      AAA Batteries (4-pack)                 3        2.99   
20459    315315      27in 4K Gaming Monitor                 1      389.99   
22046    316833             Vareebadd Phone                 1      400.00   
22081    316867            27in FHD Monitor                 1      149.99   
24614    319308        USB-C Charging Cable                 1       11.95   
24676    319366                20in Monitor                 2      109.99   

                               Purchase Address    Sales            City  \
1555            427 Wilson St, Dallas, TX 75001    14.95          Dallas   
2255       519 13th St, New York City, NY 10001   700.00   New York City   
2256       519 13th St, New York City, NY 10001    29.90   New York City   
2415        542 2nd St, San Francisco, CA 94016   600.00   San Francisco   
2796        731 7th St, New York City, NY 10001    47.96   New York City   
3537    762 Johnson St, San Francisco, CA 94016   150.00   San Francisco   
3617    754 Hickory St, New York City, NY 10001    11.95   New York City   
4914      581 River St, San Francisco, CA 94016    11.95   San Francisco   
5076     657 Spruce St, New York City, NY 10001    99.99   New York City   
5077     657 Spruce St, New York City, NY 10001    14.95   New York City   
5326               193 5th St, Austin, TX 73301   109.99          Austin   
7012         108 Jefferson St, Boston, MA 02215    14.95          Boston   
8867           825 Adams St, Portland, OR 97035     2.99        Portland   
10301            693 1st St, Portland, OR 97035   150.00        Portland   
10614     784 River St, San Francisco, CA 94016    99.99   San Francisco   
11463     431 Hickory St, Los Angeles, CA 90001   999.99     Los Angeles   
11625             3

> **The results above shows that sales for the year 2020 was made on 1st January only. But let's probe further by looking out for the unique entries for 2020 recorded sales**

In [34]:
# unique entries for the year 2020

df[df['Year'] == 2020].nunique()

Order ID            31
Product             15
Quantity Ordered     4
Price Each          15
Purchase Address    31
Sales               18
City                 8
Hour                 6
Day                  1
Month                1
Year                 1
Weekday              1
dtype: int64

> *The output above shows that the `Day`, `Month`, `Year` and `Weekday` columns have only 1 unique value confirming that all 2020 sales recorded was from Wednesday, 1st of January 2020.*

In [35]:
df[df['Year'] == 2020].count()

Order ID            34
Product             34
Quantity Ordered    34
Price Each          34
Purchase Address    34
Sales               34
City                34
Hour                34
Day                 34
Month               34
Year                34
Weekday             34
dtype: int64

> At this point, we will focus on 2019 sales data for our analysis to avoid bias.

In [36]:
# sorting out 2019 data

df = df[df['Year'] != 2020]

In [37]:
df.head()

Order ID               Product  Quantity Ordered  Price Each  \
0    295665    Macbook Pro Laptop                 1     1700.00   
1    295666    LG Washing Machine                 1      600.00   
2    295667  USB-C Charging Cable                 1       11.95   
3    295668      27in FHD Monitor                 1      149.99   
4    295669  USB-C Charging Cable                 1       11.95   

                         Purchase Address    Sales            City  Hour  Day  \
0  136 Church St, New York City, NY 10001  1700.00   New York City     0   30   
1     562 2nd St, New York City, NY 10001   600.00   New York City     7   29   
2    277 Main St, New York City, NY 10001    11.95   New York City    18   12   
3     410 6th St, San Francisco, CA 94016   149.99   San Francisco    15   22   
4           43 Hill St, Atlanta, GA 30301    11.95         Atlanta    12   18   

      Month  Year    Weekday  
0  December  2019     Monday  
1  December  2019     Sunday  
2  December  2019   Thursday  
3  December  2019     Sunday  
4  December  2019  Wednesday

In [38]:
# verifying total sales after sorting

df.groupby('Year')['Sales'].sum()

Year
2019    34483365.68
Name: Sales, dtype: float64

> As we can see from above, our data for analysis, `df` is from the year 2019 only. 

#### 2.Month with highest sales

In [39]:
# Total monthly sales

monthly_sales = df.groupby(['Month','Year'])['Sales'].sum().sort_values(ascending=False)

In [40]:
print(monthly_sales)

Month      Year
December   2019    4613443.34
October    2019    3736726.88
April      2019    3390670.24
November   2019    3199603.20
May        2019    3152606.75
March      2019    2807100.38
July       2019    2647775.76
June       2019    2577802.26
August     2019    2244467.88
February   2019    2202022.42
September  2019    2097560.13
January    2019    1813586.44
Name: Sales, dtype: float64


**`December` has the highest amount of sales with 4613443.34 and `January` has the least amount of sales with 1822256.73.**

#### 3. Most expensive and least expensive products

In [41]:
expensive_prod = df.groupby('Product')['Price Each'].max().sort_values(ascending=False)

In [42]:
print(expensive_prod)

Product
Macbook Pro Laptop            1700.00
ThinkPad Laptop                999.99
iPhone                         700.00
LG Dryer                       600.00
LG Washing Machine             600.00
Google Phone                   600.00
Vareebadd Phone                400.00
27in 4K Gaming Monitor         389.99
34in Ultrawide Monitor         379.99
Flatscreen TV                  300.00
Apple Airpods Headphones       150.00
27in FHD Monitor               149.99
20in Monitor                   109.99
Bose SoundSport Headphones      99.99
Lightning Charging Cable        14.95
Wired Headphones                11.99
USB-C Charging Cable            11.95
AA Batteries (4-pack)            3.84
AAA Batteries (4-pack)           2.99
Name: Price Each, dtype: float64


In [43]:
df['Product'].nunique()

19

**The most expensive product is `Macbook Pro Laptop` selling at `1700.00` and the least expensive product is `AAA Batteries (4-pack)` selling at `2.99`. Also there are 19 different products at sale.**

#### 4.Most purchased products

In [44]:
total_quantity_bought = df.groupby('Product')['Quantity Ordered'].sum().sort_values(ascending=False)

In [45]:
print(total_quantity_bought)

Product
AAA Batteries (4-pack)        31012
AA Batteries (4-pack)         27635
USB-C Charging Cable          23971
Lightning Charging Cable      23211
Wired Headphones              20553
Apple Airpods Headphones      15657
Bose SoundSport Headphones    13454
27in FHD Monitor               7547
iPhone                         6847
27in 4K Gaming Monitor         6243
34in Ultrawide Monitor         6199
Google Phone                   5531
Flatscreen TV                  4818
Macbook Pro Laptop             4727
ThinkPad Laptop                4128
20in Monitor                   4126
Vareebadd Phone                2067
LG Washing Machine              666
LG Dryer                        646
Name: Quantity Ordered, dtype: int64


**`AAA Batteries (4-pack)` was the most purchased with `31017` quantity sold.**

**`LG Dryer` was the least purchased with `646` quantity sold.**

#### 5. Total Quantity Purchased

In [46]:
order_qty = df['Quantity Ordered'].sum()
print(order_qty)

209038


> **Overall Quantity of products sold were `209079`.**

#### 6.City with most/least sales

In [47]:
# Total sales per City

sales_per_city = df.groupby('City')['Sales'].sum().sort_values(ascending=False)
print(sales_per_city)

City
 San Francisco    8259719.03
 Los Angeles      5451570.81
 New York City    4662976.81
 Boston           3660315.12
 Atlanta          2795498.58
 Dallas           2766260.45
 Seattle          2747605.49
 Portland         2320337.62
 Austin           1819081.77
Name: Sales, dtype: float64


**`San Francisco ` had the most sales and `Austin` had the least sales.**

#### 7. Sales per Day

In [48]:
df.head()

Order ID               Product  Quantity Ordered  Price Each  \
0    295665    Macbook Pro Laptop                 1     1700.00   
1    295666    LG Washing Machine                 1      600.00   
2    295667  USB-C Charging Cable                 1       11.95   
3    295668      27in FHD Monitor                 1      149.99   
4    295669  USB-C Charging Cable                 1       11.95   

                         Purchase Address    Sales            City  Hour  Day  \
0  136 Church St, New York City, NY 10001  1700.00   New York City     0   30   
1     562 2nd St, New York City, NY 10001   600.00   New York City     7   29   
2    277 Main St, New York City, NY 10001    11.95   New York City    18   12   
3     410 6th St, San Francisco, CA 94016   149.99   San Francisco    15   22   
4           43 Hill St, Atlanta, GA 30301    11.95         Atlanta    12   18   

      Month  Year    Weekday  
0  December  2019     Monday  
1  December  2019     Sunday  
2  December  2019   Thursday  
3  December  2019     Sunday  
4  December  2019  Wednesday

In [49]:
daily_sales = df.groupby('Weekday')['Sales'].sum().sort_values(ascending=False)

In [50]:
print(daily_sales)

Weekday
Tuesday      5087956.78
Wednesday    4980151.97
Sunday       4932169.66
Saturday     4904357.01
Monday       4883326.72
Friday       4855938.38
Thursday     4839465.16
Name: Sales, dtype: float64


**`Tuesdays` had the most sales.**

#### 7.1 Daily Sales per Month

In [51]:
daily_monthly_sales = df.groupby(['Month','Weekday'])['Sales'].sum().sort_values(ascending=False)

In [52]:
print(daily_monthly_sales)

Month      Weekday  
December   Monday       763665.71
           Tuesday      747111.03
           Sunday       738323.69
           Wednesday    619961.91
October    Tuesday      616947.93
                          ...    
September  Tuesday      265523.63
January    Friday       246617.98
           Sunday       245830.73
           Saturday     223091.95
           Monday       222134.67
Name: Sales, Length: 84, dtype: float64


#### 8. Hourly Sales

In [53]:
df.head()

Order ID               Product  Quantity Ordered  Price Each  \
0    295665    Macbook Pro Laptop                 1     1700.00   
1    295666    LG Washing Machine                 1      600.00   
2    295667  USB-C Charging Cable                 1       11.95   
3    295668      27in FHD Monitor                 1      149.99   
4    295669  USB-C Charging Cable                 1       11.95   

                         Purchase Address    Sales            City  Hour  Day  \
0  136 Church St, New York City, NY 10001  1700.00   New York City     0   30   
1     562 2nd St, New York City, NY 10001   600.00   New York City     7   29   
2    277 Main St, New York City, NY 10001    11.95   New York City    18   12   
3     410 6th St, San Francisco, CA 94016   149.99   San Francisco    15   22   
4           43 Hill St, Atlanta, GA 30301    11.95         Atlanta    12   18   

      Month  Year    Weekday  
0  December  2019     Monday  
1  December  2019     Sunday  
2  December  2019   Thursday  
3  December  2019     Sunday  
4  December  2019  Wednesday

In [64]:
hour_sales = df.groupby('Hour')['Sales'].sum().sort_values(ascending=False)

In [65]:
print(hour_sales)

Hour
19    2412938.54
12    2316821.34
11    2300610.24
20    2281716.24
18    2219348.30
13    2155389.80
17    2129361.61
14    2083672.73
21    2042000.86
10    1944286.77
15    1941549.60
16    1904601.31
9     1639030.58
22    1607549.21
8     1192348.97
23    1179304.44
7      744854.12
0      711060.59
1      458561.97
6      448113.00
2      232586.50
5      230564.88
4      162396.09
3      144697.99
Name: Sales, dtype: float64


In [58]:
# sales by product category

product_category_sales = df.groupby('Product')['Sales'].sum().sort_values(ascending=False)

In [59]:
print(product_category_sales)

Product
Macbook Pro Laptop            8035900.00
iPhone                        4792900.00
ThinkPad Laptop               4127958.72
Google Phone                  3318600.00
27in 4K Gaming Monitor        2434707.57
34in Ultrawide Monitor        2355558.01
Apple Airpods Headphones      2348550.00
Flatscreen TV                 1445400.00
Bose SoundSport Headphones    1345265.46
27in FHD Monitor              1131974.53
Vareebadd Phone                826800.00
20in Monitor                   453818.74
LG Washing Machine             399600.00
LG Dryer                       387600.00
Lightning Charging Cable       347004.45
USB-C Charging Cable           286453.45
Wired Headphones               246430.47
AA Batteries (4-pack)          106118.40
AAA Batteries (4-pack)          92725.88
Name: Sales, dtype: float64


## 4. Data Visualisation

> We are going to build a dashboard in Power BI to visualize our cleaned dataset. Let's export the dataset and load unto Tableau for the dashboard
 

In [55]:
#export data to csv format

df.to_csv('Cleaned_Sales_Data.csv', index=False)

**Link to the dashboard :** [Electronics Sales Dashboard](https://public.tableau.com/views/ElectronicsSalesDashboard_17163992269890/SalesDashboard?:language=en-US&publish=yes&:sid=&:display_count=n&:origin=viz_share_link)

## 5. Conclusions and Recommendations

In the next few lines, I will expound on the data analysis processes above and the respective findings.

First, the data I used for this project is a sales data from an electronic retailer. I then explored the data to understand it's structure and performed the neccessary cleaning. I went ahead to perform some statistics on the data to understand the story behind the data and saved the cleaned data in csv file which I loaded unto Tableau to visualize the data through a dashboard.

Below are insights generated from the data and some recommended actions:


**Data Analysis Findings**

- The data covers the year `2019`.

- The average quantity ordered is approximately `1.12`.

- The average price per item is around `$184.39`.

- The average sales amount is about `$185.48`.

- The most common hour for orders is around `3 PM (Hour 15)`.

- The total sales made over the period is `$34483365.68`.

- In January 2019, sales made was `1813586.44` and in January 2020, sales made was `8670.29`. This implies sales from January 2019 was so much higher than that of January 2020. 

- In 2019, December has the highest sales of `4613443.34` while January had the lowest sale amount of `1813586.44`.

- Of the 19 types of products sold, Macbook Pro Laptop had the highest selling price of `$1700.00` and AAA Batteries (4-pack) had the lowest selling price of `$2.99`.

- `209079` quantity of products were sold in total with AAA Batteries (4-pack) selling the most quantity at 31017 and LG Dryer being the least bought with 646 quantities sold. 

- The top-selling products by revenue are:

    - Macbook Pro Laptop: $ 8,035,900.00
    - iPhone: $ 4,792,900.00
    - ThinkPad Laptop: $ 4,127,958.72
    - Google Phone: $ 3,318,600.00
    - 27in 4K Gaming Monitor: $ 2,434,707.57 

- San Francisco generated the highest sales followed by Los Angeles and New York.

- `Tuesday` had the highest overall sales, implying that more products are bought on Tuesdays. The difference in sales amount over the week is between `$ 20,000` and `$ 58.000`.

- December generated more sales weekly as compared to the other months.

- There is a clear hourly trend in sales, with the highest sales occurring during the late morning and early evening hours. Significant fluctuations are observed throughout the day, with some hours showing much higher sales than others. The hourly sales pattern suggests that factors such as customer behavior, work schedules, and daily routines may influence sales.


**Recommendations**

- Stakeholders should invest more during the final quater of the year to attract more sales.

- AAA Batteries (4-pack) being the cheapest product, had the maximum quantity sold out. More of this product should be restocked in the shop to increase purchase from customers.

- Stakeholders should invest more in San Francisco, Los Angeles and New York. These cities should be considered for the opening of new branches.

- From the insights generated, Tuesday made the highest sales of `5087956.78` followed by Wednesday, Sunday and so on. Stakeholders can invest accordingly.

- The shop should be open and available in the late mornings and early evenings of the day to attend to all incoming orders.

- Hours with low sales can be targeted for specific marketing strategies or promotions to boost sales during those periods.

#  